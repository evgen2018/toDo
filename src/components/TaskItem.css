.fa{
    color: #333333;

}

.task-item{
    list-style-type: none;
}

.task-edit-input{
    color: #333333;
    width: 400px;
}

.task-text{
    display: inline-block;
    color: #333;
    width: 460px;
    white-space: nowrap; /* Запрещаем перенос строк */
    overflow: hidden; /* Обрезаем все, что не помещается в область */
    text-overflow: ellipsis;
}

.task-text-selected{
    color: red;
    text-decoration: line-through;
}

.btn-icon{
    cursor: pointer;
    background-color: transparent;
    border: none;
    float: right;
}

.btn-save{
    float: right;
}

.no-clickable{
    pointer-events: none;
}
