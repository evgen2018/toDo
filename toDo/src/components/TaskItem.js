import React from "react";
import PropTypes from 'prop-types';
import './TaskItem.css';
import classNames from 'classnames';

class TaskItem extends React.Component{
    static propTypes = {
        selected: PropTypes.array.isRequired,
        editTask: PropTypes.func.isRequired,
        taskData: PropTypes.shape({
            id: PropTypes.number.isRequired,
            title: PropTypes.string.isRequired,
        }).isRequired,
        deleteTask: PropTypes.func.isRequired,
        isAnyTaskEditing: PropTypes.bool.isRequired,
        toggleSelectTask: PropTypes.func.isRequired,
        toggleEditing: PropTypes.func.isRequired,
    };

    constructor(props){
        super(props);
        this.state={
            isCurrentTaskEditing:false,
            taskTitle: props.taskData.title
        };

        this.taskEditInputRef = React.createRef();
    }

    toggleEditTask = () =>{
        const {isCurrentTaskEditing} = this.state;
        this.setState({isCurrentTaskEditing: !isCurrentTaskEditing},()=>{
            if (!isCurrentTaskEditing){
                this.taskEditInputRef.current && this.taskEditInputRef.current.focus()
            }
        });

        this.props.toggleEditing();
    };

    saveEditTask = () =>{
        const {taskData, editTask} = this.props;
        editTask(taskData.id, this.state.taskTitle);
        this.toggleEditTask();
    };

    changeTaskTitle = (e) =>{
        this.setState({taskTitle: e.target.value});
    };

    toggleSelectTask = () =>{
        this.props.toggleSelectTask(this.props.taskData.id)
    };

    deleteTask = () =>{
        this.props.deleteTask(this.props.taskData.id)
    };

    render(){
        const {taskData, selected, isAnyTaskEditing} = this.props;
        const {isCurrentTaskEditing, taskTitle} = this.state;

        return (
            <li
                className={classNames(
                    'task-item',
                    isAnyTaskEditing && !isCurrentTaskEditing && 'no-clickable'
                )}>
                {isCurrentTaskEditing
                    ? <input
                        type="text"
                        className='task-edit-input'
                        ref={this.taskEditInputRef}
                        value={taskTitle}
                        onChange={this.changeTaskTitle}/>
                    : <span
                        className={classNames(
                            'task-text',
                            selected.includes(taskData.id) && 'task-text-selected'
                        )}>
                        {taskTitle}
                     </span>
                }

                {!isCurrentTaskEditing &&
                    <button
                        className='btn-icon'
                        onClick={this.deleteTask}>
                        <i className="fa fa-trash" />
                    </button>
                }
                {isCurrentTaskEditing
                    ? <button
                        className='btn btn-primary btn-save'
                        onClick={this.saveEditTask}>
                        save
                     </button>
                    : <button
                        className='btn-toggle-edit btn-icon'
                        onClick={this.toggleEditTask}>
                        <i className="fa fa-edit" />
                     </button>
                }
                {!isCurrentTaskEditing &&
                    <button
                        className='btn-icon'
                        onClick={this.toggleSelectTask}>
                        {selected.includes(taskData.id)
                            ? <i className="fa fa-minus" />
                            : <i className="fa fa-check" />}
                    </button>
                }

            </li>
        )
    }
}

export default TaskItem