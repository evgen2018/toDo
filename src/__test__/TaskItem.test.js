import React from 'react';
import TaskItem from '../components/TaskItem';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

describe('TaskItem list test', () => {
    const props = {
        taskData: {id:123, title:'mockTitle'},
        selected:[],
        isAnyTaskEditing: false,
        editTask: ()=>{},
        deleteTask: ()=>{},
        toggleEditing: ()=>{},
        toggleSelectTask: ()=>{},
    };
    const wrapper = shallow(
        <TaskItem {...props}/>
    );

    it('TaskItem renders properly', ()=>{
        expect(wrapper).toMatchSnapshot();
    });

    it('click on the btn-toggle-edit will change state isCurrentTaskEditing', ()=>{
        expect(wrapper.state().isCurrentTaskEditing).toEqual(false);
        wrapper.find('.btn-toggle-edit').simulate('click');
        expect(wrapper.state().isCurrentTaskEditing).toEqual(true);
    });

    it('input some data to task-edit-input will change state taskTitle', ()=>{
        const input = wrapper.find('.task-edit-input').first()
        expect(wrapper.state().taskTitle).toEqual('mockTitle');
        input.simulate('change', {
            target: { value: 'mockData' }
        });
        expect(wrapper.state().taskTitle).toEqual('mockData');
    });
});

