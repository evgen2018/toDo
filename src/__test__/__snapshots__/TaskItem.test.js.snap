// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem list test TaskItem renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskItem
    deleteTask={[Function]}
    editTask={[Function]}
    isAnyTaskEditing={false}
    selected={Array []}
    taskData={
      Object {
        "id": 123,
        "title": "mockTitle",
      }
    }
    toggleEditing={[Function]}
    toggleSelectTask={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span
          className="task-text"
        >
          mockTitle
        </span>,
        <button
          className="btn-icon"
          onClick={[Function]}
        >
          <i
            className="fa fa-trash"
          />
        </button>,
        <button
          className="btn-toggle-edit btn-icon"
          onClick={[Function]}
        >
          <i
            className="fa fa-edit"
          />
        </button>,
        <button
          className="btn-icon"
          onClick={[Function]}
        >
          <i
            className="fa fa-check"
          />
        </button>,
      ],
      "className": "task-item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "mockTitle",
          "className": "task-text",
        },
        "ref": null,
        "rendered": "mockTitle",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="fa fa-trash"
          />,
          "className": "btn-icon",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-trash",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="fa fa-edit"
          />,
          "className": "btn-toggle-edit btn-icon",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-edit",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <i
            className="fa fa-check"
          />,
          "className": "btn-icon",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-check",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": "button",
      },
    ],
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span
            className="task-text"
          >
            mockTitle
          </span>,
          <button
            className="btn-icon"
            onClick={[Function]}
          >
            <i
              className="fa fa-trash"
            />
          </button>,
          <button
            className="btn-toggle-edit btn-icon"
            onClick={[Function]}
          >
            <i
              className="fa fa-edit"
            />
          </button>,
          <button
            className="btn-icon"
            onClick={[Function]}
          >
            <i
              className="fa fa-check"
            />
          </button>,
        ],
        "className": "task-item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "mockTitle",
            "className": "task-text",
          },
          "ref": null,
          "rendered": "mockTitle",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fa fa-trash"
            />,
            "className": "btn-icon",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-trash",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fa fa-edit"
            />,
            "className": "btn-toggle-edit btn-icon",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-edit",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <i
              className="fa fa-check"
            />,
            "className": "btn-icon",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-check",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": "button",
        },
      ],
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
